### DISCUSSION TITLE
Can't make body parameter equality check work

### DISCUSSION THREAD

**Discussion by artem-kamyshev** [mockoon/mockoon#1216] [Help] [answered]
Hello, thanks for the great product.

I'm a bit stuck with a body parameters feature. 

I've added a rule with a body parameter and used jsonpath there (for some reason object-path failed). It works fine when I just check for its presence:

![image](https://github.com/mockoon/mockoon/assets/31185325/afc82fd0-03dd-418e-9136-a76f3a31a823)

![image](https://github.com/mockoon/mockoon/assets/31185325/bd0b7731-58ef-48fe-bfae-b15bc9b72e45)

But when I try to use equal the rule never takes effect. Can you advice me on this?

![image](https://github.com/mockoon/mockoon/assets/31185325/db73ad74-cb41-4dc8-b498-61994fd0e9a3)

![image](https://github.com/mockoon/mockoon/assets/31185325/e72d677c-a47f-47e2-b1d6-251ca8425979)

I've tried both `$.response.systime.ntp.port._text`, `$.response.systime.ntp.port`. Same results.
Here is the xml that I send in a body:

```
<?xml version="1.0" encoding="utf-8"?>
<response><systime><mode>0</mode><tz>GMT0BST,M3.5.0/1,M10.5.0</tz><datetime>2023-12-04T19:42:23</datetime><dst><active>0</active></dst><ntp><host>pool.ntp.org</host><port>123</port><interval>1</interval></ntp></systime></response>
```

Here is the corresponding part of a mockoon config:

>     {
>       "uuid": "67f37599-fd96-4d50-9826-724b50f85892",
>       "type": "http",
>       "documentation": "",
>       "method": "post",
>       "endpoint": "action/set",
>       "responses": [
>         {
>           "uuid": "d8874d7a-0574-4334-bb18-ed5e34fad9c2",
>           "body": "",
>           "latency": 0,
>           "statusCode": 200,
>           "label": "",
>           "headers": [
>             {
>               "key": "Server",
>               "value": "Brovotech/2.0.0"
>             },
>             {
>               "key": "Connection",
>               "value": "close"
>             }
>           ],
>           "bodyType": "INLINE",
>           "filePath": "",
>           "databucketID": "",
>           "sendFileAsBody": false,
>           "rules": [
>             {
>               "target": "query",
>               "modifier": "subject",
>               "value": "systime",
>               "invert": false,
>               "operator": "equals"
>             },
>             {
>               "target": "body",
>               "modifier": "$.response.systime.ntp.port._text",
>               "value": "123",
>               "invert": false,
>               "operator": "equals"
>             }
>           ],
>           "rulesOperator": "AND",
>           "disableTemplating": true,
>           "fallbackTo404": false,
>           "default": false,
>           "crudKey": "id"
>         },

Mockoon Desktop version is v5.1.0

**Comment by 255kb** [comment]
I just tested with the following call and it worked (using a `systime` query param and keeping your rule as-is). 
I think you are passing the wrong header in the request: 

![image](https://github.com/mockoon/mockoon/assets/7489814/f0c820b4-e062-46d7-9b1a-5cc189cf3347)

It should be `Content-Type: application/json` for Mockoon to parse it correctly.

**Comment by artem-kamyshev** [comment]
Thanks for the response.

It should be fine with `application/x-www-form-urlencoded` as this content-type is within a list of compatible types here: https://mockoon.com/docs/latest/route-responses/dynamic-rules/#2-property-name-or-path

Even more, it is fine when I create a new environment from a scratch:
[test.json](https://github.com/mockoon/mockoon/files/13558920/test.json)

`curl http://127.0.0.1:80/action/set -X POST -H "Content-Type: application/x-www-form-urlencoded" -d "@data.xml"`

[test_ok_log.txt](https://github.com/mockoon/mockoon/files/13558949/test_ok_log.txt)

And fails on my config:

[log_501_error.txt](https://github.com/mockoon/mockoon/files/13558966/log_501_error.txt)

501 error is my default response. It's the last in the list of responses for each request in my config.

There should be something that I'm missing in my config. I must've said from a start that I generate the config programmatically so it should be some error in it. Some kind of a wierd one as it breaks the rules even when I update them manually from Mockoon  Desktop.

I'll investigate further. I wish I had logging of a failed matches against the rules, it would make such a debugging way more easy.

**Comment by 255kb** [comment, marked as answer]
First of all, while the content-type you are using is indeed compatible with fetching data by path, the XML is not parsed if the entering request is not [`application/xml` or `text/xml`](https://mockoon.com/docs/latest/response-configuration/xml-support/). So, it's mandatory. 

Second, in your test.json example, there is only one response. In this case Mockoon will always answer with a 200 no matter what the rules are saying. 
I think the issue is really with the wrong entering Content-Type.

Try with this file: [test2.json](https://github.com/mockoon/mockoon/files/13561407/test2.json)
I created it from yours, and just added a 501 response and made it the default.
When I call it with this request, it works:
```
POST /action/set?subject=systime
Content-Type: application/xml

<?xml version="1.0" encoding="utf-8"?>
<response>
  <systime>
    <mode>0</mode>
    <tz>GMT0BST,M3.5.0/1,M10.5.0</tz>
    <datetime>2023-12-04T19:42:23</datetime>
    <dst>
      <active>0</active>
    </dst>
    <ntp>
      <host>pool.ntp.org</host>
      <port>123</port>
      <interval>1</interval>
    </ntp>
  </systime>
</response>
```

**Reply by artem-kamyshev** [reply]
Thank you for the reply. I've reproduces what you told me. That's how it works indeed. :)

Unfortunately I cannot control the headers as I use a mock-server with an external API which I cannot influence at all. I've made a fork and removed this restriction and now Mockoon does exactly what I need.
